KAFKA-9393: DeleteRecords may cause extreme lock contention for large partition directories

gardnervickers opened the pull request:
https://issues.apache.org/jira/browse/KAFKA-9393
This PR avoids a performance issue with `DeleteRecords` when a partition directory contains high numbers of files. Previously, `DeleteRecords` would iterate the partition directory searching for producer state snapshot files. With this change, the iteration is removed in favor of keeping a 1:1 mapping between producer state snapshot file and segment file. A segment files corresponding producer state snapshot file is now deleted when the segment file is deleted. 

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

ijuma commented:
Thanks for the PR. Did we consider a tweak where incrementing the log start offset doesn't trigger deletion of snapshots aside from the ones mapped to deleted segments? We delete them during checkpointing anyway.

gardnervickers commented:
@ijuma I might be misunderstanding, but with this PR, incrementing the log start offset does not trigger deletions of snapshots unless a segment/snapshot pair would be deleted due to the log start offset moving. Possibly that's different from the behavior your suggesting?

ijuma commented:
That's fine. I just mean that we don't necessarily have to keep more snapshots.

gardnervickers commented:
That could also work, but I think that keeping snapshot files around for the lifetime of their associated segments simplifies things a bit in that now snapshot files have a similar lifecycle to the indices, and also reduces the chances of us ever having to scan the log during recovery.

ijuma commented:
@hachikuji what was the original motivation for only keeping the more recent snapshots?

gardnervickers commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -870,6 +870,10 @@ class Log(@volatile var dir: File,
+    // Pass the set of loaded segment base offsets to diff against the set of producer state snapshot
+    // files. Any orphaned files should be cleaned up.
+    val keepOffsets = logSegments.map(_.baseOffset)
+    producerStateManager.cleanupOrphanSnapshotFiles(keepOffsets)
```
This should be moved to `loadLog()` since it only needs to run at startup. Also, the set of offsets to delete should exclude the log end offset, since that is snapshotted out during shutdown.

hachikuji commented:
ok to test

hachikuji commented:
@ijuma I suggested it in order to simplify some of the bookkeeping and to reduce the chance for full log scans during recovery.

hachikuji commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -310,6 +310,10 @@ class Log(@volatile var dir: File,
       throw new IllegalStateException("Producer state must be empty during log initialization")
+    // Pass the set of loaded segment base offsets to diff against the set of producer state snapshot
+    // files. Any orphaned files should be cleaned up.
+    val keepOffsets = logSegments.map(_.baseOffset)
```
nit: I'd suggest a more descriptive name like `segmentBaseOffsets`

hachikuji commented on core/src/main/scala/kafka/log/LogSegment.scala:
```diff
@@ -622,7 +623,8 @@ class LogSegment private[log] (val log: FileRecords,
       () => delete(timeIndex.deleteIfExists _, "time index", lazyTimeIndex.file, logIfMissing = true),
-      () => delete(txnIndex.deleteIfExists _, "transaction index", txnIndex.file, logIfMissing = false)
+      () => delete(txnIndex.deleteIfExists _, "transaction index", txnIndex.file, logIfMissing = false),
+      () => delete(() => {Files.deleteIfExists(producerStateSnapshot.toPath)}, "producer state snapshot", producerStateSnapshot, logIfMissing = false)
```
nit: braces are probably not needed

hachikuji commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -310,6 +310,10 @@ class Log(@volatile var dir: File,
+    // Pass the set of loaded segment base offsets to diff against the set of producer state snapshot
+    // files. Any orphaned files should be cleaned up.
+    val keepOffsets = logSegments.map(_.baseOffset)
+    producerStateManager.cleanupOrphanSnapshotFiles(keepOffsets)
```
Just mentioning this as an observation. Unlike the index files, which have a lifecycle clearly tied to the segment, producer snapshots are a bit of a mix. They are created in `Log` when a segment is rolled, but deletion is the job of the segment. I'm partially wondering if it would be clearer to handle both operations in the same context.

gardnervickers replied:
Hmm I might be misunderstanding @hachikuji but the current PR should delete producer state snapshot files along with segment deletion. I included this additional cleanup step during log opening because snapshot files are not atomically renamed as part of deletion like other files, which seems to allow for orphaned files if there is a crash during segment deletion. Are you suggesting that we handle snapshot file deletion like the deletion of index/segment files by adding the `deleted` suffix?

hachikuji replied:
What I was saying is more in line with Jun's comment below. It is more about managing the lifecycle of the snapshot consistently. Typically it's simpler if creation and deletion happen in the same context.

hachikuji commented on core/src/main/scala/kafka/log/ProducerStateManager.scala:
```diff
@@ -759,6 +759,15 @@ class ProducerStateManager(val topicPartition: TopicPartition,
+  def cleanupOrphanSnapshotFiles(keepOffsets: Iterable[Long]): Unit = {
+    val expected = keepOffsets.map(Log.producerSnapshotFile(logDir, _)).toSet ++ latestSnapshotFile
+    val actual = listSnapshotFiles.toSet
+    actual.diff(expected).foreach(file => Files.deleteIfExists(file.toPath))
```
Can we add a log message when we delete snapshot files?

hachikuji commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -310,6 +310,10 @@ class Log(@volatile var dir: File,
+    // files. Any orphaned files should be cleaned up.
+    val keepOffsets = logSegments.map(_.baseOffset)
+    producerStateManager.cleanupOrphanSnapshotFiles(keepOffsets)
     loadProducerState(logEndOffset, reloadFromCleanShutdown = hasCleanShutdownFile)
```
Was there a reason you chose to cleanup the orphaned files before loading producer state? I'm not sure it matters too much, but it does mean that we won't remove the snapshot from the log end offset until the following restart.

gardnervickers replied:
I don't have a specific reason for this, but that's a good point that we won't remove the log end offset snapshot file until the next restart. I'll swap the ordering here, thanks!

hachikuji reviewed (commented):
Thanks for the patch. I left a few small comments.

hachikuji commented:
ok to test

hachikuji commented:
retest this please

hachikuji reviewed (approved):
LGTM. Thanks for the patch!

hachikuji commented:
ok to test

hachikuji commented:
retest this please

junrao commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -311,6 +311,10 @@ class Log(@volatile var dir: File,
+    // Pass the set of loaded segment base offsets to diff against the set of producer state snapshot
+    // files. Any orphaned files should be cleaned up.
+    val segmentBaseOffsets = logSegments.map(_.baseOffset)
+    producerStateManager.cleanupOrphanSnapshotFiles(segmentBaseOffsets)
```
So, this means that we will be removing the producer snapshot corresponding to log end offset. One of the potential impact is that if the broker has an immediate hard failure after this point, the recovery will have to rebuild the producer snapshot for the active segment, which can be expensive. We probably want to think if there is any other impact.

hachikuji replied:
I think `cleanupOrphanSnapshotFiles` always retains the last snapshot.

junrao replied:
Thanks. Sounds good. We probably should change the comment above to make it clear that the last snapshot is always preserved?

junrao commented on core/src/main/scala/kafka/log/LogSegment.scala:
```diff
@@ -653,11 +655,13 @@ object LogSegment {
   def open(dir: File, baseOffset: Long, config: LogConfig, time: Time, fileAlreadyExists: Boolean = false,
            initFileSize: Int = 0, preallocate: Boolean = false, fileSuffix: String = ""): LogSegment = {
     val maxIndexSize = config.maxIndexSize
+    val producerStateSnapshot = Log.producerSnapshotFile(dir, baseOffset)
```
Hmm, should we include fileSuffix in producerStateSnapshot too?

junrao commented on core/src/main/scala/kafka/log/LogSegment.scala:
```diff
@@ -57,6 +57,7 @@ class LogSegment private[log] (val log: FileRecords,
                                val lazyOffsetIndex: LazyIndex[OffsetIndex],
                                val lazyTimeIndex: LazyIndex[TimeIndex],
                                val txnIndex: TransactionIndex,
+                               val producerStateSnapshot: File,
```
Hmm, it seems that we are missing the handling of producerStateSnapshot in a few methods like changeFileSuffixes(), updateDir(), and potentially flush().

gardnervickers replied:
Hmm I think if we want to treat producer snapshot files like other index files, it seems we may need a different approach. In this PR I was relying on `cleanupOrphanSnapshotFiles` to avoid having to deal with `changeFileSuffixes` and `updateDir`.

junrao replied:
Hmm, by putting producerStateSnapshot inside LogSegment, it seems to suggest that producerStateSnapshot should be managed as other files belonging to a segment?

ijuma replied:
I would agree.

gardnervickers replied:
That sounds good, I was just saying the approach taken in this PR will have to be rethought depending on what degree we want the `LogSegment` to manage producer state snapshots versus the `Log`/`ProducerStateManager`. I can think of two options which can be taken to improve the ownership here a bit. 

1. The `Log` continues to manage producer state snapshots via the `ProducerStateManager` and the logic for deletion is lifted back into the log layer. This means wherever we try to delete segments with `LogSegment.deleteIfExists` we also delete the corresponding producer state snapshot via a call to the `ProducerStateManager`. I think this most closely resembles the current state in trunk.

2. The `LogSegment` and `Log`/`ProducerStateManager` share ownership of the producer state snapshot with only the `LogSegment` owning deletion and access going through the `ProducerStateManager`. This may allow us to get rid of the producer state snapshot specific truncation logic in `ProducerStateManager`. The question here is what to do about producer state snapshot files which don't match a corresponding segments base offset. This may not be too bad as I think in all cases they would just be associated with the active segment.

junrao replied:
I see. The tricky thing is that the last snapshot doesn't have a corresponding log segment and needs to be managed separately outside of the LogSegment. However, for snapshots that have an associated segment, managing the snapshot file as part of the log segment seems convenient.

junrao reviewed (commented):
@gardnervickers : Thanks for the PR. The overall approach looks good. A few comments below. Also, could we test for compaction topics too? Let's say we compact segments with starting offset 100, 200, 300 into a single segment with starting offset 100. Initially, the new segment will be created with prefix .swap, which will be empty. So, it seems that we need to pass along the right snapshot file to the new segment.

gardnervickers commented:
Update: The most recent changeset attempts to more closely control the life cycle of the producer state snapshot file. Most of the changes relevant to this are in `LogSegment.scala`.

I decided that it may be overkill for this PR to create a wrapper object for encapsulating the functionality around `LogSegment::updateDir`, `LogSegment::changeFileSuffixes` and `LogSegment::deleteIfExists`, mostly because producer state snapshot files are not used like other index files. I think it's possible to work around this and restructure how producer state snapshot files are read and written, but that seems like a much larger change. 

Lastly, the issue of deletion after clean shutdown was brought up. When clean shutdown happens, we can end up with a producer snapshot file which is created at an offset for which there is no corresponding segment. I could think of two ways to deal with this:

1. Allow for one "orphan" producer state snapshot file at a time, and during clean shutdown, ensure that the previous one is deleted.
2. On startup after loading producer state, delete the "orphan" producer state snapshot file.

Both of these cases suffer a potential failure scenario where a clean shutdown is followed by an unclean shutdown and we end up having to reread part of the log, but I believe that may be ok.

I choose to implement #2 as it was simpler, and instead of just deleting the highest producer state snapshot file without a corresponding segment file, I delete all producer state snapshot files without a corresponding segment file.

junrao commented on core/src/main/scala/kafka/log/ProducerStateManager.scala:
```diff
@@ -23,7 +23,7 @@ import java.nio.file.{Files, StandardOpenOption}
 import kafka.log.Log.offsetFromFile
 import kafka.server.LogOffsetMetadata
-import kafka.utils.{Logging, nonthreadsafe, threadsafe}
+import kafka.utils.{CoreUtils, Logging, nonthreadsafe, threadsafe}
```
unused import CoreUtils

junrao commented on core/src/main/scala/kafka/log/ProducerStateManager.scala:
```diff
@@ -753,6 +753,18 @@ class ProducerStateManager(val topicPartition: TopicPartition,
+   * Remove any producer state snapshot files which do not have a corresponding offset provided
+   * in keepOffsets. The latest snapshot file will always be kept.
+   */
+  def cleanupOrphanSnapshotFiles(segmentBaseOffsets: Iterable[Long]): Unit = {
```
Since this is called on broker restart, it seems that we no longer need to call deleteSnapshotFiles() in truncateAndReload(), truncateHead() and truncate() ?

gardnervickers replied:
Thanks @junrao,

I wanted to enumerate why it's safe to remove these now to make sure we're on the same page.

- `deleteSnapshotFiles` can be deleted because snapshot file deletion is done outside of the ProducerStateManager now.

- `truncate` can be deleted because we only ever fully truncate the producer state snapshots if we are also truncating the log, and as a result, the segment files.

- `truncateAndReload` and `truncateHead` seems to be called when recovering a segment and when loading the log. Upon starting up after an unclean shutdown, we will call `truncateAndReload`. I'm a bit less certain that we can get rid of these methods, I will need to investigate further.

junrao commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -1260,7 +1265,6 @@ class Log(@volatile var dir: File,
           info(s"Incrementing log start offset to $newLogStartOffset")
           updateLogStartOffset(newLogStartOffset)
           leaderEpochCache.foreach(_.truncateFromStart(logStartOffset))
-          producerStateManager.truncateHead(newLogStartOffset)
```
We don't need to clear the producer snapshot here. However, it seems that we still need to do other things in truncateHead() like removeUnreplicatedTransactions()?

junrao commented on core/src/main/scala/kafka/log/LogSegment.scala:
```diff
@@ -47,6 +47,7 @@ import scala.math._
  * @param lazyOffsetIndex The offset index
  * @param lazyTimeIndex The timestamp index
  * @param txnIndex The transaction index
+ * @param producerStateSnapshot The producer state snapshot file matching this segments base offset
```
segments base offset => segment's base offset

junrao commented on core/src/main/scala/kafka/log/LogSegment.scala:
```diff
@@ -622,7 +631,8 @@ class LogSegment private[log] (val log: FileRecords,
       () => delete(timeIndex.deleteIfExists _, "time index", lazyTimeIndex.file, logIfMissing = true),
-      () => delete(txnIndex.deleteIfExists _, "transaction index", txnIndex.file, logIfMissing = false)
+      () => delete(txnIndex.deleteIfExists _, "transaction index", txnIndex.file, logIfMissing = false),
+      () => delete(() =>  Files.deleteIfExists(producerStateSnapshot.toPath), "producer state snapshot", producerStateSnapshot, logIfMissing = false)
```
For consistency, should we cover producerStateSnapshot in lastModified_= ?

gardnervickers replied:
Yes thanks!

junrao commented on core/src/main/scala/kafka/log/ProducerStateManager.scala:
```diff
@@ -753,6 +753,18 @@ class ProducerStateManager(val topicPartition: TopicPartition,
+   * in keepOffsets. The latest snapshot file will always be kept.
+   */
+  def cleanupOrphanSnapshotFiles(segmentBaseOffsets: Iterable[Long]): Unit = {
+    val expected = segmentBaseOffsets.map(Log.producerSnapshotFile(logDir, _)).toSet ++ latestSnapshotFile
```
Just to be clear. You mentioned 

"I choose to implement #2 as it was simpler, and instead of just deleting the highest producer state snapshot file without a corresponding segment file, I delete all producer state snapshot files without a corresponding segment file."

But, we are still keeping the snapshot file with the largest offset?

gardnervickers replied:
Yes this is a mistake, now that `cleanupOrphanSnapshotFiles` occurs after loading the producer state, we should also delete the latest snapshot file.

junrao reviewed (commented):
@gardnervickers : Thanks for the updated PR. A few more comments below.

ijuma commented:
Ok to test

ijuma commented:
ok to test

gardnervickers commented:
In 6152847, I changed the approach taken in this PR to move away from managing the producer state snapshot file as part of the segment lifecycle. This was done for two reasons, first it turned out to be a bit difficult to reason about producer state snapshot files which did not have a corresponding segment file (such as the snapshot taken during shutdown). Second, moving the management of deletion to the LogSegment ended up splitting ownership of the producer state snapshot file between the `LogSegment` and the `ProducerStateManager`, which seemed to complicate when things like truncation could happen.

Instead, a call is made to the `ProducerStateManager` to delete any matching producer snapshot files when `LogSegment`'s are deleted. This occurs during retention and as a result of log cleaning (compaction). For producer state snapshot files which do not have a corresponding segment, they will be cleaned up on restart (unless they are the latest snapshot file).

I think this ends up being simpler while more closely matching the current set of responsibilities, where the `ProducerStateManager` "owns" creating/reading/deleting snapshot files, but I'd be interested in hearing other opinions 😄.  

cc/ @junrao

ijuma commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -2237,7 +2209,11 @@ class Log(@volatile private var _dir: File,
       maybeHandleIOException(s"Error while deleting segments for $topicPartition in dir ${dir.getParent}") {
-        segments.foreach(_.deleteIfExists())
+        segments.foreach {
+          case segment =>
```
Nit: `case` and newline are unnecessary here.

junrao commented on core/src/main/scala/kafka/log/ProducerStateManager.scala:
```diff
@@ -751,6 +751,25 @@ class ProducerStateManager(val topicPartition: TopicPartition,
 
+  /**
+   * Remove any producer state snapshot files which do not have a corresponding offset provided
+   * in keepOffsets. The latest snapshot file will always be kept.
```
What's keepOffsets?

gardnervickers replied:
This should be `segmentBaseOffsets`, thanks!

junrao commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -2421,6 +2396,7 @@ class Log(@volatile private var _dir: File,
         newSegments.foreach { splitSegment =>
           splitSegment.close()
           splitSegment.deleteIfExists()
+          producerStateManager.deleteIfExists(splitSegment.baseOffset)
```
It doesn't seem that we generate producer snapshot files for those new segments?

junrao commented on core/src/main/scala/kafka/log/LogCleaner.scala:
```diff
@@ -595,8 +595,10 @@ private[log] class Cleaner(val id: Int,
-        catch {
+        try {
+          cleaned.deleteIfExists()
+          log.producerStateManager.deleteIfExists(cleaned.baseOffset)
```
Hmm, the cleaned segment has the same base offset as the first segment. So, we don't want to delete that snapshot file.

junrao commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -2237,7 +2209,10 @@ class Log(@volatile private var _dir: File,
-        segments.foreach(_.deleteIfExists())
+        segments.foreach { segment =>
+          segment.deleteIfExists()
+          producerStateManager.deleteIfExists(segment.baseOffset)
```
Hmm, this can be a bit tricky. When we replace old segments with a new segment in LogCleaner, each of the old segment will be deleted. However, the first old segment has the same offset as the new segment. So, we don't want to just delete the producer snapshot corresponding to the first old segment.

junrao reviewed (commented):
@gardnervickers : Thanks for the updated PR. A few more comments below.

ijuma commented:
ok to test

gardnervickers committed:
Rework how producer state snapshot files are maintained:

- On broker startup, only the most recent producer state snapshot and producer state snapshots associated with a segment file are retained.

- Producer state snapshot files are managed via an in-memory index inside of the ProducerStateManager to reduce lookup costs.

- Producer state snapshots are kept around for the lifetime of their associated segment to prevent large truncations from deleting all producer state snapshot files.

- Producer state snapshots are now deleted asynchronously as part of the thread which
  deletes segment and index files. Snapshot file deletion occurs when segment deletion occurs.

gardnervickers committed:
Add tests verifying that deleting stray snapshots does not remove the

largest stray snapshot.

gardnervickers committed:
The snapshots var maintained by the PoducerStateManager does not need to be volatile as it is only a set once during log initialization.

gardnervickers committed:
When deleting cleaned segment files, do not delete the
corresponding producer state snapshot file if the deleted snapshot
file is being replaced with a clean version.

gardnervickers committed:
When advancing the log start offset, ensure that the lastMapOffset is advanced in the ProducerStateManager.

gardnervickers committed:
Merge branch 'trunk' of github.com:apache/kafka into KAFKA-9393-2

gardnervickers commented:
@junrao I was able to get some more time to pick this back up, apologies for the long gap in time since my last update.

In the most recent commit, I've moved away from tracking producer state snapshot files on the segment and instead maintain an in-memory cache using the ProducerStateManager.

There were a few reasons for doing this. Firstly, it became a bit awkward trying to track the producer state snapshot file which we emit during clean shutdown since it does not really have an associated segment file. Second, the producer state truncation/restore logic would need to be moved into the `Log`, since complete view of all producer state snapshot files are required. This is further complicated by the way we handle corrupt snapshot files. Lastly, because we use a "fake" `ProducerStateManager` during segment recovery, we'd have to duplicate a lot of the same logic as exists today to handle truncation/loading outside of the segment lifecycle.

Instead, I opted to take an approach similar to the way the `Log` manages segments, by having a `ConcurrentNavigableMap` which tracks snapshot files in memory. As a result, the logic for truncation and restore largely remains the same, but instead of scanning the log directory on every operation we query the in-memory map instead. Deletions are handled in the same way that segment deletions are handled, where the snapshot file is deleted asynchronously along with the segment file. Because we scan the logdir at startup for "stray" snapshot files, it's unnecessary to rename the snapshot files pending deletion with the `.deleted` suffix.

This approach has two downsides which I think are relatively minor.

1. When a broker shuts down cleanly and emits a snapshot file, the emitted snapshot file is considered "stray" on the next broker startup. While we will clean all "stray" snapshot files except the most recent, we still keep the most recent snapshot file around until the next broker restart. This will result in a single "stray" snapshot file remaining until the next broker restart, at which point the "stray" snapshot file will be deleted.
2. Because we construct a temporary `ProducerStateManager` during segment recovery, and it may delete/create some snapshot files, we need to re-scan the log directory for snapshot files after segment loading is completed but before we load producer state. This is to ensure that the in-memory mapping for the "real" `ProducerStateManager` is up to date.

Snapshot file deletion is triggered via `Log.deleteSegmentFiles` when deletion occurs due to retention. When swapping log segments into the log (like with compaction), it appears we have the additional limitation that we don't want to delete snapshot files purely based on the base offset of the deleted segment file. To handle this case, we check to see if the segment file which is being deleted due to the swap has a counterpart new segment which is being swapped in, if it does, we do not delete the snapshot file.

gardnervickers committed:
Cleanup ProducerStateManager

- Use `snapshot` instead of `file` for snapshot name when loading producer state.
- Use `removeAndDeleteSnapshot` where applicable instead of doing it manuallly.

gardnervickers committed:
Clarify comment for LogManagerTest.

junrao commented on core/src/main/scala/kafka/log/ProducerStateManager.scala:
```diff
@@ -496,6 +491,53 @@ class ProducerStateManager(val topicPartition: TopicPartition,
+   */
+  private def loadSnapshots(): ConcurrentSkipListMap[java.lang.Long, SnapshotFile] = {
+    val tm = new ConcurrentSkipListMap[java.lang.Long, SnapshotFile]()
+    for (f <- ProducerStateManager.listSnapshotFiles(_logDir)) {
```
ProducerStateManager.listSnapshotFiles() could just be listSnapshotFiles() ?

junrao commented on core/src/main/scala/kafka/log/ProducerStateManager.scala:
```diff
@@ -496,6 +491,53 @@ class ProducerStateManager(val topicPartition: TopicPartition,
+   * corresponding to one of the provided offsets in segmentBaseOffsets will be removed, except in the case that there
+   * is a snapshot file at a higher offset than any offset in segmentBaseOffsets.
+   *
+   * The goal here is to remove any snapshot files which do not have an associated segment file, but not to remove
```
Incomplete sentence after "but not to remove".

junrao commented on core/src/main/scala/kafka/log/ProducerStateManager.scala:
```diff
@@ -496,6 +491,53 @@ class ProducerStateManager(val topicPartition: TopicPartition,
+          }
+        case None =>
+          if (!segmentBaseOffsets.contains(key)) {
+            latestStraySnapshot = Some(snapshot)
```
Could we also make sure that the offset for latestStraySnapshot is > the largest offset in segmentBaseOffsets?

gardnervickers replied:
We perform a check below which may cover this case. After setting the `snapshots` map, we look at the latest snapshot in the map. If the latest snapshot in the map is not equal to the `latestStraySnapshot`, we delete the `latestStraySnapshot`. 

I think this is a bit confusing though, so it might be better if instead we directly check that the `latestStraySnapshot` is larger than the largest offset in `segmentBaseOffsets`.

junrao commented on core/src/main/scala/kafka/log/ProducerStateManager.scala:
```diff
@@ -653,36 +697,44 @@ class ProducerStateManager(val topicPartition: TopicPartition,
     if (lastMapOffset < logStartOffset)
       lastMapOffset = logStartOffset
 
-    deleteSnapshotsBefore(logStartOffset)
```
Hmm, why don't we need to delete snapshots before logStartOffset?

gardnervickers replied:
The idea here is to clear un-replicated transactions and optionally advance the `lastMapOffset` and `lastSnapOffset` when the logStartOffset is advanced, but to leave the snapshot files around. The corresponding snapshot files should be removed during the retention pass as we cleanup the associated segment files. 

I was attempting to optimize incrementing the logStartOffset a bit so that we don't need to delete the snapshot files from the request handler thread when handling `DELETE_RECORDS`.

junrao commented on core/src/test/scala/unit/kafka/log/LogTest.scala:
```diff
@@ -1226,6 +1225,104 @@ class LogTest {
+    Files.createFile(orphanedSnapshotFile)
+    val logConfig = LogTest.createLogConfig(segmentBytes = 2048 * 5, retentionBytes = -1, fileDeleteDelayMs = 0)
+    createLog(logDir, logConfig)
+    assertEquals("expected orphaned producer state snapshot file to be cleaned up", 0, ProducerStateManager.listSnapshotFiles(logDir).size)
```
Hmm, why is orphanedSnapshotFile deleted since we keep the last stray snapshot?

gardnervickers replied:
It's being deleted because during producer state loading because we truncate producer state to match the bounds of the log, and the snapshot file written out at offset 42 is higher than the log end offset of the empty log. The test name is not very clear in this case though. 

I will fix the name and add another test which checks that we keep around the largest stray producer state snapshot file.

gardnervickers replied:
I added some extra context to the existing test and created a new test which verifies from the log that the largest stray snapshot which is within the logs end offset is retained `testLoadingLogKeepsLargestStrayProducerStateSnapshot`.

junrao commented on core/src/test/scala/unit/kafka/log/LogTest.scala:
```diff
@@ -1661,14 +1753,17 @@ class LogTest {
     log.maybeIncrementLogStartOffset(2L, ClientRecordDeletion)
+    log.deleteOldSegments() // force retention to kick in so that the snapshot files are cleaned up.
+    mockTime.sleep(logConfig.fileDeleteDelayMs + 1000) // advance the clock so file deletion takes place
+    mockTime.scheduler.tick()
```
mockTime.sleep() calls scheduler.tick() already.

junrao commented on core/src/test/scala/unit/kafka/log/ProducerStateManagerTest.scala:
```diff
@@ -834,6 +834,40 @@ class ProducerStateManagerTest {
+  @Test
+  def testRemoveAllStraySnapshots(): Unit = {
+    // Test that when stray snapshots are removed, all stray snapshots are removed when the base offset of the largest
+    // segment exceeds the offset of the largest stray snapshot.
```
Below, the base offset of the largest segment equals to and doesn't exceed the offset of the largest stray snapshot.

gardnervickers replied:
I think this sentence is a bit confusing. Snapshot 42 is not meant to be a stray snapshot here, only 5 and 2 are. I will try to reword this.

gardnervickers replied:
Hmm, I think my comment here could be worded better. Offset `42` here is not a "stray", since we provide it along with the list of segmentBaseOffsets to `removeStraySnapshots`. 

I'll change up the wording on this, thanks!

junrao reviewed (commented):
@gardnervickers : Thanks for the updated PR. A few more comments below.

junrao commented:
Also, it seems that Log.takeProducerSnapshot() no longer used?

gardnervickers committed:
Address PR feedback

- Clarify the purpose of the test which verifies truncation cleans up snapshots.

- Add new test which verifies that we delete stray snapshots below the base offset of the active segment, but keep the largest stray snapshot within the log end offset.

gardnervickers committed:
Address PR feedback

- Directly remove the latestStraySnapshot if it is less than the largest segment base offset.

gardnervickers committed:
Address PR feedback

- Reword some comments in ProducerStateManagerTest

junrao commented on core/src/test/scala/unit/kafka/log/LogTest.scala:
```diff
@@ -782,7 +782,7 @@ class LogTest {
 
     // Retain snapshots for the last 2 segments
-    ProducerStateManager.deleteSnapshotsBefore(logDir, segmentOffsets(segmentOffsets.size - 2))
+    ProducerStateManager.listSnapshotFiles(logDir).filter(_.offset < segmentOffsets(segmentOffsets.size - 2)).foreach(_.deleteIfExists())
```
Since deleteSnapshotsBefore() still exist, could we keep using it?

gardnervickers replied:
Yes, it should work if we switch these back to using deleteSnapshotsBefore. Thanks!

junrao commented on core/src/test/scala/unit/kafka/log/LogTest.scala:
```diff
@@ -794,16 +794,12 @@ class LogTest {
     // Only delete snapshots before the base offset of the recovery point segment (post KAFKA-5829 behaviour) to
     // avoid reading all segments
-    ProducerStateManager.deleteSnapshotsBefore(logDir, offsetForRecoveryPointSegment)
+    ProducerStateManager.listSnapshotFiles(logDir).filter(_.offset < offsetForRecoveryPointSegment).foreach(_.deleteIfExists())
```
Since deleteSnapshotsBefore() still exist, could we keep using it?

junrao reviewed (commented):
@gardnervickers : Thanks for the updated PR. Just a couple of minor comments below.

gardnervickers committed:
Address PR feedback

- Switch back to using `deleteSnapshotsBefore`

junrao reviewed (approved):
@gardnervickers : Thanks for the latest PR. LGTM. Waiting for the tests to pass before merging.

gardnervickers committed:
`maxOption` is not available in this version of Scala, use the explicit form instead.

gardnervickers committed:
Merge branch 'trunk' of github.com:apache/kafka into KAFKA-9393-2

gardnervickers commented:
It appears the build is running into https://github.com/apache/kafka/pull/9328.

Other than that I'm seeing two test failures in `kafka.api.MetricsTest.testMetrics` which appear to be fixed on master. I'll merge with trunk.

gardnervickers commented:
Test failures:
```
Build / JDK 11 / org.apache.kafka.connect.mirror.MirrorConnectorsIntegrationTest.testReplicationWithEmptyPartition
Build / JDK 15 / kafka.api.DelegationTokenEndToEndAuthorizationTest.testProduceConsumeViaSubscribe
Build / JDK 15 / org.apache.kafka.streams.processor.internals.StateDirectoryTest.shouldCleanupObsoleteStateDirectoriesOnlyOnce
Build / JDK 15 / org.apache.kafka.streams.processor.internals.StateDirectoryTest.shouldCleanupObsoleteStateDirectoriesOnlyOnce
Build / JDK 8 / org.apache.kafka.streams.integration.StandbyTaskEOSIntegrationTest.shouldWipeOutStandbyStateDirectoryIfCheckpointIsMissing[exactly_once]
```

I ran these locally, but it seems that these two are failing. They seem unrelated to this change though.
org.apache.kafka.streams.processor.internals.StateDirectoryTest.shouldCleanupObsoleteStateDirectoriesOnlyOnce
org.apache.kafka.streams.integration.StandbyTaskEOSIntegrationTest.shouldWipeOutStandbyStateDirectoryIfCheckpointIsMissing
